### PR TITLE
BE-DMRG Functionality and More

### PR DESCRIPTION
**Pull request by ShaunWeatherly**
*BE-DMRG*
 - Added DMRG fragment solver via interface to block2.
 - Added unittest `block2_molBE_test.py`
 - Added example script `molbe_dmrg_block2.py`
 - Added example results`../figures/BEDMRG_H8_PES20.png`

*QOL Improvements*
- Fixed grammatical mistakes in README
- Added `figures` directory to go along with example scripts
- Exposed `pyscf.lo.pipek` module to mbe
- Added `force_cleanup` feature to clear all fragment DMRG scratch files
- Changed default scratch directory for `solver=='SHCI'`
- Exposed solver-level scratch directories (`scratch`) and other solver-level kwargs (`**solver_kwargs`).

*Attributions*
- Updated reference list to be more complete and to distinguish between different aspects of the code
- Updated author attributions

### CODE REVIEW

**Comment by mscho527** [comment]
We've trying to keep the PRs bite-sized. Could you split this so that each PR adds/fixes one thing at a time? Thanks! [https://bssw.io/blog_posts/pull-request-size-matters](https://bssw.io/blog_posts/pull-request-size-matters)

**Comment by ShaunWeatherly** [comment]
As per your blog post:
>Always maximizing productivity for reviewers and the review process may create unwelcome productivity issues for other developers and collaborators submitting PRs. So far, we've focused on the productivity of the PR review process and what submitters can do to minimize review effort. However, there are productivity costs for submitters too.
>
>First, strict adherence to avoiding mixing unrelated changes in a single PR may simply discourage some good housekeeping (while I am here fixing problem A, why don't I go ahead and also fix problem B). There is just a lot of overhead associated with splitting changes over multiple PRs including creating branches and PRs, running and waiting for CI in each PR, ensuring each PR is reviewed and merged in logical order and possibly having to apply each PR's changes to multiple branches (e.g. release candidate and main). 

Also, the many small fixes will be single line PRs, and I don't think that's a valuable use of anyone's time.

**Comment by mscho527** [comment]
Your PR contains several separable parts that do not depend on each other. Separating the PR into at least three (but not 12 lol) --
- BE-DMRG related commits
- Docs improvement
- Pipek Mezey Feat

would be advantageous, especially because some of your commits are better squash-merged for clarity in the git history. The overhead you mention doesn't quite apply here, because the PRs would not be interdependent. (therefore does not require rebasing) Obviously, I can't oblige anyone to follow this, but such consideration facilitates collaborative code development.

File: README.md [hunk 1]
```diff
@@ -91,9 +88,30 @@ Comprehensive documentation for QuEmb is available at `quemb/docs`. The document
...
 - OR Meitei, T Van Voorhis, Electron correlation in 2D periodic systems, [arXiv:2308.06185](https://arxiv.org/abs/2308.06185)  
-- HZ Ye, HK Tran, T Van Voorhis, Bootstrap embedding for large molecular systems, [JCTC 16 5035 2020](https://doi.org/10.1021/acs.jctc.0c00438)
+
+### For BE-DMRG theory and implementation:
+- [INSERT BE-DMRG PAPER HERE]
+
+### For Unrestricted BE theory and implementation:
+- Tran, H. K., Ye, H.-Z. & Van Voorhis, T. Bootstrap embedding with an unrestricted mean-field bath. The Journal of Chemical Physics 153, 214101 (2020).
+- [INSERT LEAH PAPER HERE]
+
```
Comment (mscho527) [inline README.md:112]: I think (yet) unpublished papers can be added once they can be cited?

File: molbe/_opt.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def __init__(self, pot, Fobjs, Nocc, enuc,solver='MP2', ecore=0.,
                  only_chem=False, hf_veff = None,
                  hci_pt=False,hci_cutoff=0.001, ci_coeff_cutoff = None, select_cutoff=None,
                  max_space=500, conv_tol = 1.e-6,relax_density = False,
-                 ebe_hf =0.):
+                 ebe_hf =0., scratch=None, **solver_kwargs):
```
Comment (mscho527) [inline molbe/_opt.py:48]: Wouldn't it make more sense to use existing `be_var.scratch`?

File: molbe/lo.py [hunk 3]
```diff
@@ -298,6 +296,37 @@ def localize(self, lo_method, mol=None, valence_basis='sto-3g',
...
+        es_, vs_ = eigh(self.S)
+        edx = es_ > 1.e-15                
+        W_ = numpy.dot(vs_[:,edx]/numpy.sqrt(es_[edx]), vs_[:,edx].T)
+        if self.frozen_core:                    
+            P_core = numpy.eye(W_.shape[0]) - numpy.dot(self.P_core, self.S)
+            C_ = numpy.dot(P_core, W_)
+            Cpop = functools.reduce(numpy.dot,
+                                    (C_.T, self.S, C_))
+            Cpop = numpy.diag(Cpop)
+            no_core_idx = numpy.where(Cpop > 0.55)[0]
```
Comment (mscho527) [inline molbe/lo.py:314]: Maybe I'm missing something, but where does the 0.55 value come from?

File: molbe/solver.py [hunk 4]
```diff
@@ -139,7 +145,15 @@ def be_func(pot, Fobjs, Nocc, solver, enuc, hf_veff=None,
...
-        
+
+            if scratch is None and be_var.CREATE_SCRATCH_DIR:
+                tmp = os.path.join(be_var.SCRATCH, str(os.getpid()), str(fobj.dname))
+            elif scratch is None:
+                tmp = be_var.SCRATCH
+            else:
+                tmp = os.path.join(scratch, str(os.getpid()), str(fobj.dname))
+            if not os.path.isdir(tmp):
+                os.system('mkdir -p '+tmp)
```
Comment (mscho527) [inline molbe/solver.py:156]: `tmp` isn't referenced afterwards, is it?

**Review by mscho527 — CHANGES_REQUESTED** [review]

**Comment by mscho527** [comment]
`pytest` fails for the added DMRG test. Please check and modify the code as necessary.

File: molbe/_opt.py (hunk 2 above)
Comment (ShaunWeatherly) [reply molbe/_opt.py:48]: It does indeed default to be_var.SCRATCH? All this kwarg accomplishes is to expose the fragment-level scratch directories to the user if desired.

File: molbe/lo.py (hunk 3 above)
Comment (ShaunWeatherly) [reply molbe/lo.py:314]: This is just the cutoff used elsewhere in the .lo module for defining occupieds. See below for more examples.

File: molbe/solver.py (hunk 4 above)
Comment (ShaunWeatherly) [reply molbe/solver.py:156]: `tmp` is used by the DMRG solver in place of scratch to allow for easy removal afterwards if `force_cleanup=True`. This is a ham-fisted way of doing cleanup, but it ensures no unintended files are removed in the process.

File: README.md (hunk 1 above)
Comment (ShaunWeatherly) [reply README.md:112]: Indeed, these are placeholders that will ensure proper attributions are made in the future.

**Comment by ShaunWeatherly** [comment]
This is an import error, which has reminded me that the `dmrgscf` module is not a standard import for `pyscf`. As unit tests shouldn't depend on optional imports, I'm thinking it's probably best to remove this test entirely.

**Comment by ShaunWeatherly** [comment]
> Your PR contains several separable parts that do not depend on each other. Separating the PR into at least three (but not 12 lol) --
> 
> * BE-DMRG related commits
> * Docs improvement
> * Pipek Mezey Feat
> 
> would be advantageous, especially because some of your commits are better squash-merged for clarity in the git history. The overhead you mention doesn't quite apply here, because the PRs would not be interdependent. (therefore does not require rebasing) Obviously, I can't oblige anyone to follow this, but such consideration facilitates collaborative code development.

I'll apologize but maintain that these changes are related, other than the docs which would comprise very minor PRs. 

Moreover, the purpose of splitting PRs is to distribute work among different reviewers, but as you're currently the only one of us with permission to merge I fail to see how the extra work on my part will will aid in this task. 

Finally, I think whatever standards we agree to should be agreed upon as a group, as this is a codebase that the group will build on, and this requires at the very least an open discussion of these topics rather than one person making these decisions seemingly behind closed doors. Respectfully, I think such considerations would actually facilitate collaborative code development.

**Comment by mscho527** [comment]
Just so you know, I am not the only one with the permission to merge. Bite-sizing the PRs was agreed amongst the people who contributed in the past months. I am applying the same standards here.

File: README.md (hunk 1 above)
Comment (oimeitei) [reply README.md:112]: This should probably come as a completely separate PR! List of BE papers can be included in the documentation  but README is probably not the best place. The references here are for QuEmb library.

Comment (ShaunWeatherly) [reply README.md:112]: That's a fair point, I'll open a few new PRs in the coming days with all of this reorganized.

**Comment by oimeitei** [comment]
Bite-sizing PRs are not only to ease review tasks but maintains better version control and help in isolating issues later on. This is more or less a standard practice everywhere. I am closing this PRs for now.

**Pull request closed without merging**